src/chain/shared_state.rs

```diff
@@ -0,0 +1,174 @@
+    pub fn our_info(&self) -> &SectionInfo {
+        // TODO: Replace `our_infos` with a new `NonemptyVec` type that statically guarantees that
+        // it's never empty.
+        unwrap!(self.opt_our_info())
+    }
+
+    pub fn our_prefix(&self) -> &Prefix<XorName> {
+        self.opt_our_info()
+            .map_or(&DEFAULT_PREFIX, SectionInfo::prefix)
+    }
+
+    pub fn our_version(&self) -> u64 {
+        self.opt_our_info().map_or(0, |si| *si.version())
+    }
+
+    /// Returns our section info with the given hash, if it exists.
+    pub fn our_info_by_hash(&self, hash: &Digest256) -> Option<&SectionInfo> {
+        self.our_infos
+            .iter()
+            .find(|(sec_info, _)| sec_info.hash() == hash)
+            .map(|(sec_info, _)| sec_info)
+    }
+
+    /// Returns the section info matching our own name with the given version number.
+    pub fn our_info_by_version(&self, version: u64) -> Option<&SectionInfo> {
+        // TODO: Binary search? Reverse order? Benchmark which one is fastest in practice.
+        self.our_infos
+            .iter()
+            .find(|(sec_info, _)| *sec_info.version() == version)
+            .map(|(sec_info, _)| sec_info)
+    }
+
+    /// Remove our infos that are older than the given version.
+    pub(super) fn clean_our_infos(&mut self, oldest_version: u64) {
+        // TODO since the list is sorted according to the version, choose something more efficient
+        self.our_infos
+            .retain(|(si, _)| *si.version() >= oldest_version)
+    }
+
+    /// Returns `true` if we have accumulated self `NetworkEvent::OurMerge`.
+    pub(super) fn is_self_merge_ready(&self) -> bool {
+        self.merging.contains(self.our_info().hash())
+    }
+
+    /// Returns the next section info if both we and our sibling have signalled for merging.
+    pub(super) fn try_merge(
+        &mut self,
+        their_info: &SectionInfo,
+    ) -> Result<Option<SectionInfo>, RoutingError> {
+        let our_hash = *self.our_info().hash();
+        let their_hash = their_info.hash();
+
+        if self.merging.contains(their_hash) && self.merging.contains(&our_hash) {
+            let _ = self.merging.remove(their_hash);
+            let _ = self.merging.remove(&our_hash);
+            self.new_info = self.our_info().merge(their_info)?;
+            Ok(Some(self.new_info.clone()))
+        } else {
+            Ok(None)
+        }
+    }
+
+    /// Returns `true` if we should merge.
+    pub(super) fn should_vote_for_merge<'a, I>(
+        &self,
+        min_section_size: usize,
+        neighbour_infos: I,
+    ) -> bool
+    where
+        I: IntoIterator<Item = &'a SectionInfo>,
+    {
+        let pfx = self.our_prefix();
+        if pfx.is_empty() || self.change == SectionChange::Splitting {
+            return false;
+        }
+
+        if self.our_info().members().len() < min_section_size {
+            return true;
+        }
+
+        for si in neighbour_infos {
```

Comment:
Not 100% sure about moving from any algo to a for loop help clarify things.

Reply:
Matter of taste, but I think `for` is cleaner for simple loops. Anyway, it's questionable in this case, so I'll revert it to `any`.